### PR TITLE
Add/updated creds flow calypso

### PR DESCRIPTION
**Pull request by vianasw**
#### Changes proposed in this Pull Request

* This patch implements stream processing of `application/x-ndjson responses`. It is a port of the code we used in the `wpcom-xhr-request` library (see [PR here](https://github.com/Automattic/wpcom-xhr-request/pull/31)). We've also brought the advanced-credentials component from jetpack-cloud.

#### Testing instructions

* To test this PR you'll need to apply this diff in your wpcom sandbox: D75408-code.

* Then, change your `/etc/hosts` file to point to your wpcom sandbox.

* Start calypso locally: `yarn start`

* Create a JN site and add a Backup plan to it.

* Then go to http://calypso.localhost:3000/settings/jetpack/<JN_site_url>#credentials

* And test the credentials flow with valid and invalid credentials. It should behave the same way as in jetpack-cloud.

### CODE REVIEW

File: client/components/advanced-credentials/credentials-form/style.scss [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+		font-size: $font-body-small;
+	}
+
+	@include breakpoint-deprecated( '>660px' ) {
+		.credentials-form__intro-text {
+			font-size: $font-body;
+		}
+	}
+
+	&__buttons .credentials-form__buttons-flex {
```
Comment (jsnajdr) [inline client/components/advanced-credentials/credentials-form/style.scss]: Let's write this as expanded `.credentials-form__buttons`. The `&__buttons` shortcut makes it very difficult to grep for usages of the CSS class name.

File: client/components/advanced-credentials/connection-status.tsx [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	Connecting,
+	Failed,
+	Loading,
+}
+
+interface Props {
+	state: StatusState;
+}
+
+const ConnectionStatus: FunctionComponent< Props > = ( { state } ) => {
```
Comment (jsnajdr) [inline client/components/advanced-credentials/connection-status.tsx]: Are you duplicating all the `advanced-credentials` Jetpack Cloud components here? There should be no need for the duplication, just move the files to the common directory and update existing imports.

File: client/my-sites/site-settings/settings-jetpack/controller.js [hunk 3]
```diff
@@ -2,6 +2,6 @@ import { createElement } from 'react';
 import JetpackMain from 'calypso/my-sites/site-settings/settings-jetpack/main';
 
 export function jetpack( context, next ) {
-	context.primary = createElement( JetpackMain );
+	context.primary = createElement( JetpackMain, { context } );
```
Comment (jsnajdr) [inline client/my-sites/site-settings/settings-jetpack/controller.js]: Instead of passing the entire `context` object as a prop, can you pass just the `query`?

Also, any component can read the current route information with the `getCurrentQueryArguments` Redux selector.

**Comment by nunyvega** [comment]
Thanks for your review @jsnajdr !
I modified the code to use `getCurrentQueryArguments` and expanded CSS instead of using `&`. 
We'll now check if it's possible to use the same component on JP Cloud and here.

**Comment by nunyvega** [comment]
We modified Calypso Green to use the shared component and removed the duplicate code.

File: client/components/advanced-credentials/host-info.ts [hunk 4]
```diff
@@ -16,23 +15,23 @@ enum InfoTypes {
 
 export interface Text {
 	type: InfoTypes.Text;
-	text: ReactChild;
+	text: string;
```
Comment (jsnajdr) [inline client/components/advanced-credentials/host-info.ts]: You don't want to do any changes in this file. The `ReactChild` type is correct, as the text can be a React markup (with HTML), not just a string. And there shouldn't be any `translate( ... ).toString()`, just `translate()`.

File: client/components/advanced-credentials/credentials-form/index.tsx [hunk 5]
```diff
@@ -50,7 +50,7 @@ const ServerCredentialsForm: FunctionComponent< Props > = ( {
 					...formState,
 					protocol: currentTarget.value as 'ftp' | 'ssh',
 					port: ( () => {
-						let port = parseInt( String( formState.port ) );
```
Comment (jsnajdr) [inline client/components/advanced-credentials/credentials-form/index.tsx]: What's the justification for removing `String()` here?

File: client/components/advanced-credentials/verification/index.tsx [hunk 6]
```diff
@@ -18,8 +18,6 @@ interface Props {
 	onReview: () => void;
 }
 
-const ERROR_CODES_FOR_BLOCKED_REQUEST = [ 'service_unavailable', 'invalid_credentials' ];
-
```
Comment (jsnajdr) [inline client/components/advanced-credentials/verification/index.tsx]: Is this change relevant to this PR? It removes special error reporting for these two errors.

File: client/my-sites/site-settings/settings-jetpack/main.jsx [hunk 7]
```diff
@@ -1,23 +1,27 @@
...
+import AdvancedCredentials from 'calypso/components/advanced-credentials';
 import DocumentHead from 'calypso/components/data/document-head';
 import QueryRewindState from 'calypso/components/data/query-rewind-state';
 import QuerySitePurchases from 'calypso/components/data/query-site-purchases';
 import EmptyContent from 'calypso/components/empty-content';
 import FormattedHeader from 'calypso/components/formatted-header';
 import Main from 'calypso/components/main';
 import SidebarNavigation from 'calypso/my-sites/sidebar-navigation';
-import JetpackCredentials from 'calypso/my-sites/site-settings/jetpack-credentials';
+//import JetpackCredentials from 'calypso/my-sites/site-settings/jetpack-credentials';
```
Comment (jsnajdr) [inline client/my-sites/site-settings/settings-jetpack/main.jsx]: This import should be removed rather than commented out.

File: client/my-sites/site-settings/settings-jetpack/main.jsx [hunk 8]
```diff
@@ -1,23 +1,27 @@
...
 import SiteSettingsNavigation from 'calypso/my-sites/site-settings/navigation';
 import { siteHasScanProductPurchase } from 'calypso/state/purchases/selectors';
+import getCurrentQueryArguments from 'calypso/state/selectors/get-current-query-arguments';
 import isRewindActive from 'calypso/state/selectors/is-rewind-active';
 import isSiteFailedMigrationSource from 'calypso/state/selectors/is-site-failed-migration-source';
 import { isJetpackSite } from 'calypso/state/sites/selectors';
 import { getSelectedSite, getSelectedSiteId } from 'calypso/state/ui/selectors';
 
 const SiteSettingsJetpack = ( { site, siteId, siteIsJetpack, showCredentials, translate } ) => {
+	const { host, action } = useSelector( getCurrentQueryArguments );
```
Comment (jsnajdr) [inline client/my-sites/site-settings/settings-jetpack/main.jsx]: Let's not mix `connect` and `useSelector` in one component. Either add a new selector call to the `connect` map function, or convert the entire component to hooks.

**Comment by nunyvega** [comment]
I was mistakenly overwriting the latest changes on this code, which was removing `ERROR_CODES_FOR_BLOCKED_REQUEST` and removing `String` from `let port`, basically reverting the code to the previous version of it. Thanks for catching that, and for the thorough review @jsnajdr !
I corrected that and added the query arguments to the `connect` map function.

File: packages/wpcom-proxy-request/src/index.js [hunk 9]
```diff
@@ -18,6 +18,8 @@ const proxyOrigin = 'https://public-api.wordpress.com';
  */
 const origin = window.location.protocol + '//' + window.location.host;
 
+let onStreamRecord = null;
```
Comment (jsnajdr) [inline packages/wpcom-proxy-request/src/index.js:21]: This is a global variable, but it would be shared by all requests running in parallel. It's better to store the callback as `xhr.onStreamRecord = ...`, then store the `xhr` in `requests[ id ]`, and then retrieve it before called:
```js
if ( statusCode === 207 ) {
  const onStreamRecord = xhr.onStreamRecord;
  if ( onStreamRecord ) {
    onStreamRecord( body );
  }
}
```

**Review by jsnajdr — COMMENTED** [review]
There's one bug in `wpcom-proxy-request`, the rest of PR is OK :+1: Thanks for addressing all feedback!
